source/module_io/test/read_wfc_to_rho_test.cpp

```diff
@@ -0,0 +1,290 @@
+            for (int ib = 0; ib < nbands; ++ib)
+            {
+                ofs << "  " << ib + 1 << "                  0.0000000                " << 1.0 << std::endl;
+            }
+            ofs << std::endl;
+        }
+        ofs.close();
+    }
+
+    // Init Psi
+    psi = new psi::Psi<std::complex<double>>(nks, nbands, wfcpw->npwk_max, wfcpw->npwk);
+    std::complex<double>* ptr = psi->get_pointer();
+    for (int i = 0; i < nks * nbands * wfcpw->npwk_max; i++)
+    {
+        ptr[i] = std::complex<double>((i + my_pool * 100) / 100.0, (i + my_pool) / 100.0);
+    }
+
+    // Init charge
+    chg.rho = new double*[1];
+    chg._space_rho = new double[rhopw->nrxx];
+    chg.rho[0] = chg._space_rho;
+    ModuleBase::GlobalFunc::ZEROS(chg.rho[0], rhopw->nrxx);
+    chg.rhopw = rhopw;
+    chg.nrxx = rhopw->nrxx;
+    // set charge_ref
+    Charge chg_ref;
+    chg_ref.rho = new double*[1];
+    chg_ref._space_rho = new double[rhopw->nrxx];
+    chg_ref.rho[0] = chg_ref._space_rho;
+    ModuleBase::GlobalFunc::ZEROS(chg_ref.rho[0], rhopw->nrxx);
+    std::vector<std::complex<double>> rho_tmp(rhopw->nrxx);
+    chg_ref.nrxx = rhopw->nrxx;
+
+    for (int ik = 0; ik < nks; ++ik)
+    {
+        for (int ib = 0; ib < nbands; ++ib)
+        {
+            const std::complex<double>* wfc_ib = ptr + ik * nbands * wfcpw->npwk_max + ib * wfcpw->npwk_max;
+            wfcpw->recip2real(wfc_ib, rho_tmp.data(), ik);
+
+            const double w1 = wg(ik, ib) / wfcpw->omega;
+
+            for (int ir = 0; ir < rhopw->nrxx; ir++)
+            {
+                chg_ref.rho[0][ir] += w1 * std::norm(rho_tmp[ir]);
+            }
+        }
+    }
+
+#ifdef __MPI
+    chg_ref.init_chgmpi();
+    chg_ref.reduce_diff_pools(chg_ref.rho[0]);
+#endif
+
+    // Write the wave functions to file
+    ModuleIO::write_wfc_pw("WAVEFUNC", *psi, *kv, wfcpw);
+
+    // Read the wave functions to charge density
+    ModuleIO::read_wfc_to_rho(wfcpw, nkstot, kv->isk, chg);
+
+    // compare the charge density
+    for (int ir = 0; ir < rhopw->nrxx; ++ir)
+    {
+        EXPECT_NEAR(chg.rho[0][ir], chg_ref.rho[0][ir], 1e-8);
+    }
+    // std::cout.precision(16);
+    // std::cout<<chg.rho[0][0]<<std::endl;
+    if (GlobalV::NPROC == 1)
+        EXPECT_NEAR(chg.rho[0][0], 8617.076357957576, 1e-8);
+    else if (GlobalV::NPROC == 4)
+    {
+        const std::vector<double> ref = {8207.849135313403, 35.34776105132742, 8207.849135313403, 35.34776105132742};
+        EXPECT_NEAR(chg.rho[0][0], ref[GlobalV::MY_RANK], 1e-8);
+        // for (int ip = 0; ip < GlobalV::NPROC; ++ip)
+        // {
+        //     if (GlobalV::MY_RANK == ip)
+        //     {
+        //         std::cout.precision(16);
+        //         std::cout << GlobalV::MY_RANK << " " << chg.rho[0][0] << std::endl;
+        //     }
+        //     MPI_Barrier(MPI_COMM_WORLD);
+        // }
+    }
+
+    delete[] chg.rho;
```

Comment:
try using std::vector?

Reply:
Rho is a member of Charge class. It is a double** pointer.